### PR TITLE
Add concurrent execution support to Executor

### PR DESCRIPTION
**Pull request by wilhuff**

### CODE REVIEW

File: Firestore/core/test/firebase/firestore/util/executor_test.cc [hunk 1]
```diff
@@ -243,6 +244,60 @@ TEST_P(ExecutorTest, DuplicateTagsOnOperationsAreAllowed) {
...
+    int count() const {
+      std::lock_guard<std::mutex> lock(mutex_);
+      return threads_;
+    }
+
+    /** Awaits the completion of all threads. */
+    void Await() {
+      std::unique_lock<std::mutex> lock(mutex_);
+      while (threads_ > 0) {
+        is_zero_.wait(lock);
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/util/executor_test.cc]: Why not check the number of threads in the callback given to `wait`?

File: Firestore/core/test/firebase/firestore/util/executor_test.cc [hunk 2]
```diff
@@ -243,6 +244,60 @@ TEST_P(ExecutorTest, DuplicateTagsOnOperationsAreAllowed) {
...
+
+    /**
+     * Bumps the counter down by one and waits for the counter to become zero.
+     */
+    void Bump() {
+      std::unique_lock<std::mutex> lock(mutex_);
+
+      threads_ -= 1;
+
+      is_zero_.wait(lock, [this] { return threads_ == 0; });
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/util/executor_test.cc]: Is this to ensure concurrency (guarantee that `Bump`s cannot all happen on a single thread)? Can you add a comment?

File: Firestore/core/test/firebase/firestore/util/executor_test.cc [hunk 3]
```diff
@@ -243,6 +244,60 @@ TEST_P(ExecutorTest, DuplicateTagsOnOperationsAreAllowed) {
...
+      is_zero_.notify_all();
+    }
+
+   private:
+    mutable std::mutex mutex_;
+    std::condition_variable is_zero_;
+    int threads_ = 0;
+  };
+
+  const int threads = 5;
```
Comment (var-const) [inline Firestore/core/test/firebase/firestore/util/executor_test.cc]: Would it be worthwhile to check the case where there's more/less tasks than threads? (i.e., check with `4` and `6` threads)

File: Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm [hunk 4]
```diff
@@ -342,6 +342,14 @@ void MarkDone() {
   return absl::make_unique<ExecutorLibdispatch>(queue);
 }
 
+std::unique_ptr<Executor> Executor::CreateConcurrent(const char* label, int) {
+  // Concurrent queues auto-create enough threads to avoid deadlock so there's
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/executor_libdispatch.mm]: Now that you mention it, I think that's probably right. `ExecutorStd` uses exactly the number of threads specified, so `CreateConcurrent("foo", 1)` is equivalent to `CreateSerial("foo")`. We wouldn't want callers to accidentally come to rely on this equivalence.

File: Firestore/core/src/firebase/firestore/util/executor_std.cc [hunk 5]
```diff
@@ -38,23 +38,34 @@ std::string ThreadIdToString(const std::thread::id thread_id) {
...
 
 ExecutorStd::~ExecutorStd() {
   shutting_down_ = true;
-  // Make sure the worker thread is not blocked, so that the call to `join`
-  // doesn't hang.
-  UnblockQueue();
-  worker_thread_.join();
+
+  // Make sure the worker threads are not blocked, so that the call to `join`
+  // doesn't hang. It's not deterministic which threads will pick up the entries
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/executor_std.cc]: Reworked.

File: Firestore/core/src/firebase/firestore/util/executor_std.cc [hunk 6]
```diff
@@ -38,23 +38,34 @@ std::string ThreadIdToString(const std::thread::id thread_id) {
...
+  // so an entry for all threads before attempting to join.
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/executor_std.cc]: Done.

File: Firestore/core/src/firebase/firestore/util/executor_std.cc [hunk 7]
```diff
@@ -38,23 +38,34 @@ std::string ThreadIdToString(const std::thread::id thread_id) {
...
+  for (size_t i = 0; i < worker_thread_pool_.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/executor_std.cc]: I had that initially, but the compiler gave a warning for the unused variable.

File: Firestore/core/src/firebase/firestore/util/executor_std.cc [hunk 8]
```diff
@@ -116,15 +127,25 @@ ExecutorStd::Id ExecutorStd::NextId() {
...
+  if (IsCurrentExecutor()) {
+    return Name();
+  } else {
+    return ThreadIdToString(std::this_thread::get_id());
+  }
 }
 
 std::string ExecutorStd::Name() const {
-  return ThreadIdToString(worker_thread_.get_id());
+  return ThreadIdToString(worker_thread_pool_[0].get_id());
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/util/executor_std.cc]: Done.

File: Firestore/core/test/firebase/firestore/util/executor_libdispatch_test.mm [hunk 9]
```diff
@@ -30,9 +30,14 @@
 
 using testutil::Expectation;
 
-std::unique_ptr<Executor> ExecutorFactory() {
-  return absl::make_unique<ExecutorLibdispatch>(
-      dispatch_queue_create("ExecutorLibdispatchTests", DISPATCH_QUEUE_SERIAL));
+std::unique_ptr<Executor> ExecutorFactory(int threads = 1) {
+  if (threads == 1) {
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_libdispatch_test.mm]: Done.

File: Firestore/core/test/firebase/firestore/util/executor_test.cc (hunk 1 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: I meant to fix these to be the same before sending this. Sorry about that.

File: Firestore/core/test/firebase/firestore/util/executor_test.cc [hunk 10]
```diff
@@ -243,6 +244,60 @@ TEST_P(ExecutorTest, DuplicateTagsOnOperationsAreAllowed) {
...
+      }
+    }
+
+    /**
+     * Bumps the counter down by one and waits for the counter to become zero.
+     */
+    void Bump() {
+      std::unique_lock<std::mutex> lock(mutex_);
+
+      threads_ -= 1;
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: Done.

File: Firestore/core/test/firebase/firestore/util/executor_test.cc (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: Yes; the class comment tries to do this, but apparently not enough. Added a comment here too.

File: Firestore/core/test/firebase/firestore/util/executor_test.cc [hunk 11]
```diff
@@ -243,6 +244,60 @@ TEST_P(ExecutorTest, DuplicateTagsOnOperationsAreAllowed) {
...
+
+      is_zero_.wait(lock, [this] { return threads_ == 0; });
+
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: Done.

File: Firestore/core/test/firebase/firestore/util/executor_test.cc [hunk 12]
```diff
@@ -243,6 +244,60 @@ TEST_P(ExecutorTest, DuplicateTagsOnOperationsAreAllowed) {
...
+      is_zero_.notify_all();
+    }
+
+   private:
+    mutable std::mutex mutex_;
+    std::condition_variable is_zero_;
+    int threads_ = 0;
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: Done.

File: Firestore/core/test/firebase/firestore/util/executor_test.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: Done.

Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: Checking with more tasks than there are backing threads would deadlock. Checking with fewer tasks than the number of available threads doesn't really test anything different than what this is testing so I'm disinclined to add the test.

File: Firestore/core/test/firebase/firestore/util/executor_test.cc [hunk 13]
```diff
@@ -243,6 +244,60 @@ TEST_P(ExecutorTest, DuplicateTagsOnOperationsAreAllowed) {
...
+  const int threads = 5;
+  executor = GetParam()(threads);
+  auto countdown = std::make_shared<BlockingCountdown>(threads);
+
+  for (int i = 0; i < threads; i++) {
+    executor->Execute([countdown] { countdown->Bump(); });
+  }
+
+  countdown->Await();
+  ASSERT_EQ(0, countdown->count());
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: junit still uses yoda order and while this particular test doesn't have to be portable to Android, it's helpful to keep these aligned. A constant source of friction previously on iOS was that junit and xctest had the reverse conventions and porting tests between the platforms required reversing these.

[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/test/firebase/firestore/util/executor_test.cc]: Cool, thanks for explaining.

**Comment by wilhuff** [comment]
To answer your earlier question: yes, the tests [pass under TSAN](https://travis-ci.org/firebase/firebase-ios-sdk/jobs/611657766?utm_medium=notification&utm_source=github_status).

**Pull request merged**